Process.status() returns 'running' for a ZombieProcess

This is on a Mac OSX El Capitan Here is a quick session to demonstrate. Process 70 on my machine right now is indeed a zombie process.

```
Python 2.7.10 (default, Jun 10 2015, 19:43:32)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import psutil
>>> psutil.__version__
'4.0.0'
>>> p = psutil.Process(pid=70)
>>> p.status()
'running'
>>> p.name()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gigi/.virtualenvs/work/lib/python2.7/site-packages/psutil/__init__.py", line 564, in name
    cmdline = self.cmdline()
  File "/Users/gigi/.virtualenvs/work/lib/python2.7/site-packages/psutil/__init__.py", line 617, in cmdline
    return self._proc.cmdline()
  File "/Users/gigi/.virtualenvs/work/lib/python2.7/site-packages/psutil/_psosx.py", line 207, in wrapper
    raise ZombieProcess(self.pid, self._name, self._ppid)
psutil.ZombieProcess: psutil.ZombieProcess process still exists but it's a zombie (pid=70)
```

I followed the code to the status() method in _psxosx.py to see why it returns 'running'

```
    @wrap_exceptions
    def status(self):
        code = cext.proc_status(self.pid)
        # XXX is '?' legit? (we're not supposed to return it anyway)
        return PROC_STATUSES.get(code, '?')
```

The cext.proc_status(self.pid) returns 2, which is the code for 'running'.

Many tests currently fail immediately OS X 10.11 El Capitain.

I'd recommend adding a Travis CI target for OSX.

```
======================================================================
FAIL: test__all__ (test_misc.TestMisc)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_misc.py", line 156, in test__all__
    self.fail('%r not in psutil.__all__' % name)
AssertionError: 'x' not in psutil.__all__

======================================================================
FAIL: test_vmem_free (test_osx.TestSystemAPIs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/__init__.py", line 494, in wrapper
    raise err
AssertionError: 68771840L != 27615232 within 512000 delta

======================================================================
FAIL: test_fetch_all (test_process.TestFetchAllProcesses)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1594, in test_fetch_all
    self.fail(''.join(failures))
AssertionError:
    ======================================================================
    FAIL: test_exe (proc=psutil.Process(pid=82273, name='cupsd'), ret='/usr/sbin/cupsd'))
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1579, in test_fetch_all
        meth(ret, p)
      File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1614, in exe
        self.assertTrue(os.access(ret, os.X_OK))
      File "/Users/riggle/.pyenv/versions/2.7.9/lib/python2.7/unittest/case.py", line 422, in assertTrue
        raise self.failureException(msg)
    AssertionError: False is not true


======================================================================
FAIL: test_proc_cwd (test_process.TestNonUnicode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 2021, in test_proc_cwd
    self.assertEqual(encode_path(p.cwd()), funny_directory)
AssertionError: '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/tmptYaTwJ/%C0%80' != '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/tmptYaTwJ/\xc0\x80'

======================================================================
FAIL: test_proc_exe (test_process.TestNonUnicode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1983, in test_proc_exe
    self.assertEqual(encode_path(os.path.basename(p.exe())), b"\xc0\x80")
AssertionError: '%C0%80' != '\xc0\x80'

======================================================================
FAIL: test_proc_open_files (test_process.TestNonUnicode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 2036, in test_proc_open_files
    self.assertIn(funny_file, encode_path(path))
AssertionError: '/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/tmptYaTwJ/\xc0\x80' not found in '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/tmptYaTwJ/%C0%80'

======================================================================
FAIL: test_prog_w_funky_name (test_process.TestProcess)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 744, in test_prog_w_funky_name
    os.path.normcase(funky_path))
AssertionError: '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psu1GIghwfoo bar )' != '/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psu1GIghwfoo bar )'

======================================================================
FAIL: test_proc_cwd (test_process.TestUnicode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1905, in test_proc_cwd
    self.assertEqual(p.cwd(), tdir)
AssertionError: '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psutil-e\xcc\x80-RaOftt' != '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psutil-\xc3\xa8-RaOftt'

======================================================================
FAIL: test_proc_open_files (test_process.TestUnicode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1919, in test_proc_open_files
    self.assertEqual(os.path.normcase(path), os.path.normcase(self.uexe))
AssertionError: '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psuExQYoPe\xcc\x80' != '/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psuExQYoP\xc3\xa8'

----------------------------------------------------------------------
Ran 359 tests in 4.451s

FAILED (failures=9, skipped=195)
make: *** [test] Error 1
```

Nope, the test failures are still present during `make test`, and the Zombie process issue also repros.

```
$ python --version
Python 3.5.1
$ python -c 'import psutil; print(psutil.__version__)'
4.2.0
$ uname -a
Darwin riggle-macbookpro2 15.4.0 Darwin Kernel Version 15.4.0: Fri Feb 26 22:08:05 PST 2016; root:xnu-3248.40.184~3/RELEASE_X86_64 x86_64
$ make test
======================================================================
ERROR: test_halfway_terminated_process (test_process.TestProcess)
----------------------------------------------------------------------
RuntimeError: proc_pidinfo(PROC_PIDLISTFDS) failed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1277, in test_halfway_terminated_process
    ret = meth()
  File "/Users/riggle/psutil/psutil/__init__.py", line 1081, in open_files
    return self._proc.open_files()
  File "/Users/riggle/psutil/psutil/_psosx.py", line 208, in wrapper
    return fun(self, *args, **kwargs)
  File "/Users/riggle/psutil/psutil/_psosx.py", line 313, in open_files
    rawlist = cext.proc_open_files(self.pid)
SystemError: <class 'OSError'> returned a result with an error set

======================================================================
ERROR: test_zombie_process (test_process.TestProcess)
----------------------------------------------------------------------
RuntimeError: proc_pidinfo(PROC_PIDLISTFDS) failed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1342, in test_zombie_process
    zproc.as_dict()
  File "/Users/riggle/psutil/psutil/__init__.py", line 473, in as_dict
    ret = attr()
  File "/Users/riggle/psutil/psutil/__init__.py", line 1081, in open_files
    return self._proc.open_files()
  File "/Users/riggle/psutil/psutil/_psosx.py", line 208, in wrapper
    return fun(self, *args, **kwargs)
  File "/Users/riggle/psutil/psutil/_psosx.py", line 313, in open_files
    rawlist = cext.proc_open_files(self.pid)
SystemError: <class 'OSError'> returned a result with an error set

======================================================================
FAIL: test_vmem_free (test_osx.TestSystemAPIs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/__init__.py", line 494, in wrapper
    raise err
  File "/Users/riggle/psutil/psutil/tests/__init__.py", line 491, in wrapper
    return fun(*args, **kwargs)
  File "/Users/riggle/psutil/psutil/tests/test_osx.py", line 156, in test_vmem_free
    delta=MEMORY_TOLERANCE)
AssertionError: 3845955584 != 1574113280 within 512000 delta

======================================================================
FAIL: test_fetch_all (test_process.TestFetchAllProcesses)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1594, in test_fetch_all
    self.fail(''.join(failures))
AssertionError:
    ======================================================================
    FAIL: test_open_files (proc=psutil.Process(pid=1, name='launchd'))
    ----------------------------------------------------------------------
    RuntimeError: proc_pidinfo(PROC_PIDLISTFDS) failed

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1559, in test_fetch_all
        ret = attr(*args)
      File "/Users/riggle/psutil/psutil/__init__.py", line 1081, in open_files
        return self._proc.open_files()
      File "/Users/riggle/psutil/psutil/_psosx.py", line 208, in wrapper
        return fun(self, *args, **kwargs)
      File "/Users/riggle/psutil/psutil/_psosx.py", line 313, in open_files
        rawlist = cext.proc_open_files(self.pid)
    SystemError: <class 'OSError'> returned a result with an error set


======================================================================
FAIL: test_proc_cwd (test_process.TestNonUnicode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 2021, in test_proc_cwd
    self.assertEqual(encode_path(p.cwd()), funny_directory)
AssertionError: b'/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/tmp5gyeiu5j/%C0%80' != b'/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/tmp5gyeiu5j/\xc0\x80'

======================================================================
FAIL: test_proc_exe (test_process.TestNonUnicode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1983, in test_proc_exe
    self.assertEqual(encode_path(os.path.basename(p.exe())), b"\xc0\x80")
AssertionError: b'%C0%80' != b'\xc0\x80'

======================================================================
FAIL: test_proc_open_files (test_process.TestNonUnicode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 2036, in test_proc_open_files
    self.assertIn(funny_file, encode_path(path))
AssertionError: b'/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/tmp5gyeiu5j/\xc0\x80' not found in b'/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/tmp5gyeiu5j/%C0%80'

======================================================================
FAIL: test_prog_w_funky_name (test_process.TestProcess)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 744, in test_prog_w_funky_name
    os.path.normcase(funky_path))
AssertionError: '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psu8reo_y0wfoo bar )' != '/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psu8reo_y0wfoo bar )'
- /private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psu8reo_y0wfoo bar )
? --------
+ /var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psu8reo_y0wfoo bar )


======================================================================
FAIL: test_proc_cwd (test_process.TestUnicode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1905, in test_proc_cwd
    self.assertEqual(p.cwd(), tdir)
AssertionError: '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psutil-è-x40mmfk_' != '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psutil-è-x40mmfk_'
- /private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psutil-è-x40mmfk_
?                                                                 ^^
+ /private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psutil-è-x40mmfk_
?                                                                 ^


======================================================================
FAIL: test_proc_open_files (test_process.TestUnicode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/riggle/psutil/psutil/tests/test_process.py", line 1919, in test_proc_open_files
    self.assertEqual(os.path.normcase(path), os.path.normcase(self.uexe))
AssertionError: '/private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psux5cme4_xè' != '/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psux5cme4_xè'
- /private/var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psux5cme4_xè
? --------                                                            ^^
+ /var/folders/hv/wnvzfp2d73b8cdxj7w_98tv0008rd2/T/psux5cme4_xè
?                                                             ^


----------------------------------------------------------------------
Ran 359 tests in 3.912s

FAILED (failures=8, errors=2, skipped=195)
...

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -14,6 +14,8 @@ Bug tracker at https://github.com/giampaolo/psutil/issues
 
 **Bug fixes**
 
+- #783: [OSX] Process.status() may erroneously return "running" for zombie
+  processes.
 - #798: [Windows] Process.open_files() returns and empty list on Windows 10.
 - #825: [Linux] cpu_affinity; fix possible double close and use of unopened
   socket.
@@ -25,6 +27,9 @@ Bug tracker at https://github.com/giampaolo/psutil/issues
 - #906: [BSD] disk_partitions(all=False) returned an empty list. Now the
   argument is ignored and all partitions are always returned.
 - #907: [FreeBSD] Process.exe() may fail with OSError(ENOENT).
+- #908: [OSX, BSD] different process methods could errounesuly mask the real
+  error for high-privileged PIDs and raise NoSuchProcess and AccessDenied
+  instead of OSError and RuntimeError.
 
 
 4.3.1 - 2016-09-01

--- a/psutil/_psutil_bsd.c
+++ b/psutil/_psutil_bsd.c
@@ -59,6 +59,8 @@
 #include <netinet/in.h>   // process open files/connections
 #include <sys/un.h>
 
+#include "_psutil_common.h"
+
 #ifdef __FreeBSD__
     #include "arch/bsd/freebsd.h"
     #include "arch/bsd/freebsd_socks.h"

--- a/psutil/_psutil_bsd.c
+++ b/psutil/_psutil_bsd.c
@@ -559,9 +559,10 @@ psutil_proc_open_files(PyObject *self, PyObject *args) {
     if (psutil_kinfo_proc(pid, &kipp) == -1)
         goto error;
 
+    errno = 0;
     freep = kinfo_getfile(pid, &cnt);
     if (freep == NULL) {
-        psutil_raise_ad_or_nsp(pid);
+        psutil_raise_for_pid(pid, "kinfo_getfile() failed");
         goto error;
     }
 

--- a/psutil/_psutil_common.c
+++ b/psutil/_psutil_common.c
@@ -88,15 +88,27 @@ psutil_pid_exists(long pid) {
 }
 
 
+/*
+ * Utility used for those syscalls which do not return a meaningful
+ * error that we can translate into an exception which makes sense.
+ * As such, we'll have to guess.
+ * On UNIX, if errno is set, we return that one (OSError).
+ * Else, if PID does not exist we assume the syscall failed because
+ * of that so we raise NoSuchProcess.
+ * If none of this is true we giveup and raise RuntimeError(msg).
+ * This will always set a Python exception and return NULL.
+ */
 int
-psutil_raise_ad_or_nsp(long pid) {
+psutil_raise_for_pid(long pid, char *msg) {
     // Set exception to AccessDenied if pid exists else NoSuchProcess.
-    int ret;
-    ret = psutil_pid_exists(pid);
-    if (ret == 0)
+    if (errno != 0) {
+        PyErr_SetFromErrno(PyExc_OSError);
+        return 0;
+    }
+    if (psutil_pid_exists(pid) == 0)
         NoSuchProcess();
-    else if (ret == 1)
-        AccessDenied();
-    return ret;
+    else
+        PyErr_SetString(PyExc_RuntimeError, msg);
+    return 0;
 }
 #endif

--- a/psutil/_psutil_common.c
+++ b/psutil/_psutil_common.c
@@ -6,6 +6,11 @@
  * Routines common to all platforms.
  */
 
+#ifdef PSUTIL_POSIX
+#include <sys/types.h>
+#include <signal.h>
+#endif
+
 #include <Python.h>
 
 
@@ -35,3 +40,63 @@ AccessDenied(void) {
     Py_XDECREF(exc);
     return NULL;
 }
+
+
+#ifdef PSUTIL_POSIX
+/*
+ * Check if PID exists. Return values:
+ * 1: exists
+ * 0: does not exist
+ * -1: error (Python exception is set)
+ */
+int
+psutil_pid_exists(long pid) {
+    int ret;
+
+    // No negative PID exists, plus -1 is an alias for sending signal
+    // too all processes except system ones. Not what we want.
+    if (pid < 0)
+        return 0;
+
+    // As per "man 2 kill" PID 0 is an alias for sending the seignal to
+    // every  process in the process group of the calling process.
+    // Not what we want.
+    if (pid == 0) {
+#if defined(PSUTIL_LINUX) || defined(BSD)
+        // PID 0 does not exist at leas on Linux and all BSDs.
+        return 0;
+#else
+        // On OSX it does.
+        // TODO: check Solaris.
+        return 1;
+#endif
+    }
+
+    ret = kill(pid , 0);
+    if (ret == 0)
+        return 1;
+    else {
+        if (errno == ESRCH)
+            return 0;
+        else if (errno == EPERM)
+            return 1;
+        else {
+            PyErr_SetFromErrno(PyExc_OSError);
+            return -1;
+        }
+    }
+}
+
+
+int
+psutil_raise_ad_or_nsp(long pid) {
+    // Set exception to AccessDenied if pid exists else NoSuchProcess.
+    int ret;
+    ret = psutil_pid_exists(pid);
+    if (ret == 0)
+        NoSuchProcess();
+    else if (ret == 1)
+        AccessDenied();
+    return ret;
+}
+#endif

--- a/psutil/_psutil_common.c
+++ b/psutil/_psutil_common.c
@@ -62,12 +62,10 @@ psutil_pid_exists(long pid) {
     // every  process in the process group of the calling process.
     // Not what we want.
     if (pid == 0) {
-#if defined(PSUTIL_LINUX) || defined(BSD)
-        // PID 0 does not exist at leas on Linux and all BSDs.
+#if defined(PSUTIL_LINUX) || defined(PSUTIL_FREEBSD)
+        // PID 0 does not exist on these platforms.
         return 0;
 #else
-        // On OSX it does.
-        // TODO: check Solaris.
         return 1;
 #endif
     }

--- a/psutil/_psutil_common.h
+++ b/psutil/_psutil_common.h
@@ -11,5 +11,5 @@ PyObject* NoSuchProcess(void);
 
 #ifdef PSUTIL_POSIX
 int psutil_pid_exists(long pid);
-int psutil_raise_ad_or_nsp(long pid);
+void psutil_raise_for_pid(long pid, char *msg);
 #endif

--- a/psutil/_psutil_common.h
+++ b/psutil/_psutil_common.h
@@ -8,3 +8,8 @@
 
 PyObject* AccessDenied(void);
 PyObject* NoSuchProcess(void);
+
+#ifdef PSUTIL_POSIX
+int psutil_pid_exists(long pid);
+int psutil_raise_ad_or_nsp(long pid);
+#endif

--- a/psutil/_psutil_osx.c
+++ b/psutil/_psutil_osx.c
@@ -169,9 +169,10 @@ psutil_proc_exe(PyObject *self, PyObject *args) {
 
     if (! PyArg_ParseTuple(args, "l", &pid))
         return NULL;
+    errno = 0;
     ret = proc_pidpath(pid, &buf, sizeof(buf));
     if (ret == 0) {
-        psutil_raise_ad_or_nsp(pid);
+        psutil_raise_for_pid(pid, "proc_pidpath() syscall failed");
         return NULL;
     }
 #if PY_MAJOR_VERSION >= 3
@@ -309,9 +310,11 @@ psutil_proc_memory_maps(PyObject *self, PyObject *args) {
         goto error;
 
     err = task_for_pid(mach_task_self(), pid, &task);
-
     if (err != KERN_SUCCESS) {
-        psutil_raise_ad_or_nsp(pid);
+        if (psutil_pid_exists(pid) == 0)
+            NoSuchProcess();
+        else
+            AccessDenied();
         goto error;
     }
 
@@ -578,7 +581,10 @@ psutil_proc_memory_uss(PyObject *self, PyObject *args) {
 
     err = task_for_pid(mach_task_self(), pid, &task);
     if (err != KERN_SUCCESS) {
-        psutil_raise_ad_or_nsp(pid);
+        if (psutil_pid_exists(pid) == 0)
+            NoSuchProcess();
+        else
+            AccessDenied();
         return NULL;
     }
 
@@ -1025,7 +1031,10 @@ psutil_proc_threads(PyObject *self, PyObject *args) {
     // task_for_pid() requires special privileges
     err = task_for_pid(mach_task_self(), pid, &task);
     if (err != KERN_SUCCESS) {
-        psutil_raise_ad_or_nsp(pid);
+        if (psutil_pid_exists(pid) == 0)
+            NoSuchProcess();
+        else
+            AccessDenied();
         goto error;
     }
 

--- a/psutil/_psutil_osx.c
+++ b/psutil/_psutil_osx.c
@@ -67,20 +67,6 @@ psutil_sys_vminfo(vm_statistics_data_t *vmstat) {
 
 
 /*
- * Set exception to AccessDenied if pid exists else NoSuchProcess.
- */
-void
-psutil_raise_ad_or_nsp(long pid) {
-    int ret;
-    ret = psutil_pid_exists(pid);
-    if (ret == 0)
-        NoSuchProcess();
-    else
-        AccessDenied();
-}
-
-
-/*
  * Return a Python list of all the PIDs running on the system.
  */
 static PyObject *

--- a/psutil/arch/bsd/freebsd.c
+++ b/psutil/arch/bsd/freebsd.c
@@ -66,51 +66,6 @@ psutil_kinfo_proc(const pid_t pid, struct kinfo_proc *proc) {
 }
 
 
-/*
- * Return 1 if PID exists in the current process list, else 0, -1
- * on error.
- * TODO: this should live in _psutil_posix.c but for some reason if I
- * move it there I get a "include undefined symbol" error.
- */
-int
-psutil_pid_exists(long pid) {
-    int ret;
-
-    if (pid < 0)
-        return 0;
-    if (pid == 0)
-        return 1;
-
-    ret = kill(pid , 0);
-    if (ret == 0)
-        return 1;
-    else {
-        if (errno == ESRCH)
-            return 0;
-        else if (errno == EPERM)
-            return 1;
-        else {
-            PyErr_SetFromErrno(PyExc_OSError);
-            return -1;
-        }
-    }
-}
-
-
-
-int
-psutil_raise_ad_or_nsp(long pid) {
-    // Set exception to AccessDenied if pid exists else NoSuchProcess.
-    int ret;
-    ret = psutil_pid_exists(pid);
-    if (ret == 0)
-        NoSuchProcess();
-    else if (ret == 1)
-        AccessDenied();
-    return ret;
-}
-
-
 // remove spaces from string
 static void psutil_remove_spaces(char *str) {
     char *p1 = str;

--- a/psutil/arch/bsd/freebsd.c
+++ b/psutil/arch/bsd/freebsd.c
@@ -559,9 +559,10 @@ psutil_proc_cwd(PyObject *self, PyObject *args) {
     if (psutil_kinfo_proc(pid, &kipp) == -1)
         goto error;
 
+    errno = 0;
     freep = kinfo_getfile(pid, &cnt);
     if (freep == NULL) {
-        psutil_raise_ad_or_nsp(pid);
+        psutil_raise_for_pid(pid, "kinfo_getfile() failed");
         goto error;
     }
 
@@ -611,9 +612,10 @@ psutil_proc_num_fds(PyObject *self, PyObject *args) {
     if (psutil_kinfo_proc(pid, &kipp) == -1)
         return NULL;
 
+    errno = 0;
     freep = kinfo_getfile(pid, &cnt);
     if (freep == NULL) {
-        psutil_raise_ad_or_nsp(pid);
+        psutil_raise_for_pid(pid, "kinfo_getfile() failed");
         return NULL;
     }
     free(freep);
@@ -777,9 +779,10 @@ psutil_proc_memory_maps(PyObject *self, PyObject *args) {
     if (psutil_kinfo_proc(pid, &kp) == -1)
         goto error;
 
+    errno = 0;
     freep = kinfo_getvmmap(pid, &cnt);
     if (freep == NULL) {
-        psutil_raise_ad_or_nsp(pid);
+        psutil_raise_for_pid(pid, "kinfo_getvmmap() failed");
         goto error;
     }
     for (i = 0; i < cnt; i++) {

--- a/psutil/arch/bsd/freebsd.h
+++ b/psutil/arch/bsd/freebsd.h
@@ -10,8 +10,6 @@ typedef struct kinfo_proc kinfo_proc;
 
 int psutil_get_proc_list(struct kinfo_proc **procList, size_t *procCount);
 int psutil_kinfo_proc(const pid_t pid, struct kinfo_proc *proc);
-int psutil_pid_exists(long pid);
-int psutil_raise_ad_or_nsp(long pid);
 
 //
 PyObject* psutil_cpu_count_phys(PyObject* self, PyObject* args);

--- a/psutil/arch/bsd/freebsd_socks.c
+++ b/psutil/arch/bsd/freebsd_socks.c
@@ -25,7 +25,7 @@
 #include <net/if_media.h>
 #include <libutil.h>
 
-#include "freebsd.h"
+#include "../../_psutil_common.h"
 
 
 #define HASHSIZE 1009

--- a/psutil/arch/bsd/freebsd_socks.c
+++ b/psutil/arch/bsd/freebsd_socks.c
@@ -497,9 +497,10 @@ psutil_proc_connections(PyObject *self, PyObject *args) {
         goto error;
     }
 
+    errno = 0;
     freep = kinfo_getfile(pid, &cnt);
     if (freep == NULL) {
-        psutil_raise_ad_or_nsp(pid);
+        psutil_raise_for_pid(pid, "kinfo_getfile() failed");
         goto error;
     }
 

--- a/psutil/arch/bsd/netbsd.c
+++ b/psutil/arch/bsd/netbsd.c
@@ -39,7 +39,6 @@
 #include <arpa/inet.h>
 
 
-#include "netbsd.h"
 #include "netbsd_socks.h"
 #include "../../_psutil_common.h"
 
@@ -50,16 +49,6 @@
 // ============================================================================
 // Utility functions
 // ============================================================================
-
-
-int
-psutil_raise_ad_or_nsp(long pid) {
-    // Set exception to AccessDenied if pid exists else NoSuchProcess.
-    if (psutil_pid_exists(pid) == 0)
-        NoSuchProcess();
-    else
-        AccessDenied();
-}
 
 
 int
@@ -121,31 +110,6 @@ kinfo_getfile(pid_t pid, int* cnt) {
 
     *cnt = (int)(len / sizeof(struct kinfo_file));
     return kf;
-}
-
-
-int
-psutil_pid_exists(pid_t pid) {
-    // Return 1 if PID exists in the current process list, else 0, -1
-    // on error.
-    // TODO: this should live in _psutil_posix.c but for some reason if I
-    // move it there I get a "include undefined symbol" error.
-    int ret;
-    if (pid < 0)
-        return 0;
-    ret = kill(pid , 0);
-    if (ret == 0)
-        return 1;
-    else {
-        if (ret == ESRCH)
-            return 0;
-        else if (ret == EPERM)
-            return 1;
-        else {
-            PyErr_SetFromErrno(PyExc_OSError);
-            return -1;
-        }
-    }
 }
 
 

--- a/psutil/arch/bsd/netbsd.c
+++ b/psutil/arch/bsd/netbsd.c
@@ -513,9 +513,10 @@ psutil_proc_num_fds(PyObject *self, PyObject *args) {
     if (! PyArg_ParseTuple(args, "l", &pid))
         return NULL;
 
+    errno = 0;
     freep = kinfo_getfile(pid, &cnt);
     if (freep == NULL) {
-        psutil_raise_ad_or_nsp(pid);
+        psutil_raise_for_pid(pid, "kinfo_getfile() failed");
         return NULL;
     }
     free(freep);

--- a/psutil/arch/bsd/netbsd.h
+++ b/psutil/arch/bsd/netbsd.h
@@ -13,11 +13,9 @@ int psutil_kinfo_proc(pid_t pid, kinfo_proc *proc);
 struct kinfo_file * kinfo_getfile(pid_t pid, int* cnt);
 int psutil_get_proc_list(kinfo_proc **procList, size_t *procCount);
 char *psutil_get_cmd_args(pid_t pid, size_t *argsize);
-PyObject * psutil_get_cmdline(pid_t pid);
-int psutil_pid_exists(pid_t pid);
-int psutil_raise_ad_or_nsp(long pid);
 
 //
+PyObject *psutil_get_cmdline(pid_t pid);
 PyObject *psutil_proc_threads(PyObject *self, PyObject *args);
 PyObject *psutil_virtual_mem(PyObject *self, PyObject *args);
 PyObject *psutil_swap_mem(PyObject *self, PyObject *args);

--- a/psutil/arch/bsd/openbsd.c
+++ b/psutil/arch/bsd/openbsd.c
@@ -36,7 +36,6 @@
 #include <err.h> // for warn() & err()
 
 
-#include "openbsd.h"
 #include "../../_psutil_common.h"
 
 #define PSUTIL_KPT2DOUBLE(t) (t ## _sec + t ## _usec / 1000000.0)
@@ -109,42 +108,6 @@ kinfo_getfile(long pid, int* cnt) {
 
     *cnt = (int)(len / sizeof(struct kinfo_file));
     return kf;
-}
-
-
-int
-psutil_pid_exists(long pid) {
-    // Return 1 if PID exists in the current process list, else 0, -1
-    // on error.
-    // TODO: this should live in _psutil_posix.c but for some reason if I
-    // move it there I get a "include undefined symbol" error.
-    int ret;
-    if (pid < 0)
-        return 0;
-    ret = kill(pid , 0);
-    if (ret == 0)
-        return 1;
-    else {
-        if (ret == ESRCH)
-            return 0;
-        else if (ret == EPERM)
-            return 1;
-        else {
-            PyErr_SetFromErrno(PyExc_OSError);
-            return -1;
-        }
-    }
-}
-
-
-int
-psutil_raise_ad_or_nsp(long pid) {
-    // Set exception to AccessDenied if pid exists else NoSuchProcess.
-    if (psutil_pid_exists(pid) == 0)
-        NoSuchProcess();
-    else
-        AccessDenied();
-    return 0;
 }
 
 

--- a/psutil/arch/bsd/openbsd.c
+++ b/psutil/arch/bsd/openbsd.c
@@ -404,9 +404,10 @@ psutil_proc_num_fds(PyObject *self, PyObject *args) {
     if (psutil_kinfo_proc(pid, &kipp) == -1)
         return NULL;
 
+    errno = 0;
     freep = kinfo_getfile(pid, &cnt);
     if (freep == NULL) {
-        psutil_raise_ad_or_nsp(pid);
+        psutil_raise_for_pid(pid, "kinfo_getfile() failed");
         return NULL;
     }
     free(freep);
@@ -505,9 +506,10 @@ psutil_proc_connections(PyObject *self, PyObject *args) {
         goto error;
     }
 
+    errno = 0;
     freep = kinfo_getfile(pid, &cnt);
     if (freep == NULL) {
-        psutil_raise_ad_or_nsp(pid);
+        psutil_raise_for_pid(pid, "kinfo_getfile() failed");
         goto error;
     }
 

--- a/psutil/arch/bsd/openbsd.h
+++ b/psutil/arch/bsd/openbsd.h
@@ -14,8 +14,6 @@ struct kinfo_file * kinfo_getfile(long pid, int* cnt);
 int psutil_get_proc_list(struct kinfo_proc **procList, size_t *procCount);
 char **_psutil_get_argv(long pid);
 PyObject * psutil_get_cmdline(long pid);
-int psutil_pid_exists(long pid);
-int psutil_raise_ad_or_nsp(long pid);
 
 //
 PyObject *psutil_proc_threads(PyObject *self, PyObject *args);

--- a/psutil/arch/osx/process_info.c
+++ b/psutil/arch/osx/process_info.c
@@ -24,38 +24,6 @@
 
 
 /*
- * Return 1 if PID exists in the current process list, else 0, -1
- * on error.
- * TODO: this should live in _psutil_posix.c but for some reason if I
- * move it there I get a "include undefined symbol" error.
- */
-int
-psutil_pid_exists(long pid) {
-    int ret;
-    if (pid < 0)
-        return 0;
-    ret = kill(pid , 0);
-    if (ret == 0)
-        return 1;
-    else {
-        return 0;
-        /*
-        // This is how it is handled on other POSIX systems but it causes
-        // test_halfway_terminated test to fail with AccessDenied.
-        if (ret == ESRCH)
-            return 0;
-        else if (ret == EPERM)
-            return 1;
-        else {
-            PyErr_SetFromErrno(PyExc_OSError);
-            return -1;
-        }
-        */
-    }
-}
-
-
-/*
  * Returns a list of all BSD processes on the system.  This routine
  * allocates the list and puts it in *procList and a count of the
  * number of entries in *procCount.  You are responsible for freeing

--- a/psutil/arch/osx/process_info.h
+++ b/psutil/arch/osx/process_info.h
@@ -11,7 +11,6 @@ typedef struct kinfo_proc kinfo_proc;
 int psutil_get_argmax(void);
 int psutil_get_kinfo_proc(pid_t pid, struct kinfo_proc *kp);
 int psutil_get_proc_list(kinfo_proc **procList, size_t *procCount);
-int psutil_pid_exists(long pid);
 int psutil_proc_pidinfo(long pid, int flavor, void *pti, int size);
 PyObject* psutil_get_cmdline(long pid);
 PyObject* psutil_get_environ(long pid);

